File: cpp/13.roman_to_integer.cpp
#include <map>
#include <string>
class Solution {
public:
    int romanToInt(std::string s) {
        std::map<char, int> roman;
        roman['I'] = 1;
        roman['V'] = 5;
        roman['X'] = 10;
        roman['L'] = 50;
        roman['C'] = 100;
        roman['D'] = 500;
        roman['M'] = 1000;
        int ans = 0;
        for (int i = 0; i < s.length(); i++) {
            ans += roman[s[i]];
            if (i > 0) {
                if (s[i - 1] == 'I' && (s[i] == 'V' || s[i] == 'X')) ans -= 2;
                if (s[i - 1] == 'X' && (s[i] == 'L' || s[i] == 'C')) ans -= 20;
                if (s[i - 1] == 'C' && (s[i] == 'D' || s[i] == 'M')) ans -= 200;
            }
        }
        return ans;
    }
};

File: py/13.roman_to_integer.py
class Solution:
    def romanToInt(self, s: str) -> int:
        roman = {'I': 1,
                 'V': 5,
                 'X': 10,
                 'L': 50,
                 'C': 100,
                 'D': 500,
                 'M': 1000}
        ans = 0
        for i, value in enumerate(s):
            ans += roman[value]
            if i > 0:
                if s[i - 1] == 'I' and (s[i] == 'V' or s[i] == 'X'):
                    ans -= 2
                if s[i - 1] == 'X' and (s[i] == 'L' or s[i] == 'C'):
                    ans -= 20
                if s[i - 1] == 'C' and (s[i] == 'D' or s[i] == 'M'):
                    ans -= 200
        return ans
